### PR TITLE
SPARK-9545, SPARK-9547: Use Maven in PRB if title contains "[test-maven]"

### PR DESCRIPTION
**Pull request by pwendell**
This is just some small glue code to actually make use of the
AMPLAB_JENKINS_BUILD_TOOL switch. As far as I can tell, we actually
don't currently use the Maven support in the tool even though it exists.
This patch switches to Maven when the PR title contains "test-maven".

There are a few small other pieces of cleanup in the patch as well.

### CODE REVIEW

**Comment by rxin** [comment]
Can it be in the pull request description or the comments?

**Comment by pwendell** [comment]
@rxin unfortunately neither of those is supported by the current PRB infrastructure. We only have access to a limited number of contextual variables set by the PRB plugin, and neither the description or comments are available.

**Comment by pwendell** [comment]
@rxin I've also just added support for triggering different hadoop version builds.

File: dev/run-tests.py [hunk 1]
```diff
@@ -231,7 +231,7 @@ def exec_maven(mvn_args=()):
     """Will call Maven in the current directory with the list of mvn_args passed
     in and returns the subprocess for any further processing"""
 
-    run_cmd([os.path.join(SPARK_HOME, "build", "mvn")] + mvn_args)
+    run_cmd([os.path.join(SPARK_HOME, "build", "mvn")] + ["--force"] + mvn_args)
```
Comment (JoshRosen) [inline dev/run-tests.py]: Technically, the `--force` can be in the first list, but it doesn't really matter.

**Comment by JoshRosen** [comment]
Just so that you're aware: the diff-dependent test selection logic is not currently implemented for Maven, so pull requests which specify Maven testing will cause all tests to be run.

**Comment by pwendell** [comment]
Jenkins, retest this please.

File: dev/run-tests.py [hunk 2]
```diff
@@ -283,16 +284,35 @@ def get_hadoop_profiles(hadoop_version):
...
+
+
+def kill_zinc_on_port(zinc_port):
+    """
+    Kill the Zinc process running on the given port, if one exists.
+    """
+
+    cmd = ("/usr/sbin/lsof -P |grep %s | grep LISTEN "
+           "| awk '{ print $2; }' | xargs kill") % zinc_port
+    # TODO: Not sure what happens here if no process exists
```
Comment (JoshRosen) [inline dev/run-tests.py]: On a non-zero exit code, `run_cmd` will print an error message and will call `sys.exit()`, so we may want to ensure that this always returns a zero exit code.  If this is a concern, you could always just perform the underlying `subprocess` call yourself.

**Comment by pwendell** [comment]
Jenkins, retest this please.

**Comment by pwendell** [comment]
Jenkins, retest this please.

**Comment by pwendell** [comment]
Jenkins, test this please.

**Comment by pwendell** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by pwendell** [comment]
Jenkins, retest this please.

**Comment by pwendell** [comment]
Jenkins, test this please.

**Comment by pwendell** [comment]
Jenkins, test this please.

**Comment by pwendell** [comment]
Jenkins, test this please.

**Comment by pwendell** [comment]
Jenkins, test this please.

**Comment by pwendell** [comment]
Okay this seems to be passing now - any thoughts @JoshRosen or @vanzin? IMO this would be really nice since we can test build changes with either build, before they make it into spark.

**Comment by vanzin** [comment]
I think this is a good initial step; like Reynold, I'd prefer if this was read from the PR's description. Perhaps the code can be amended later to get the description using the github API and parsing it (and then you can get fancy with how to declare test parameters too).

Also, it seems like the script to merge prs won't clean up these test directives, right? That's more stuff to remember when merging PRs...

On the nits side:
- your description says "maven-test" while the directive is "test-maven"
- the actual code seems to match "test-maven", not "[test-maven]" as the PR title suggests.

But overall I'm fine with it, we can keep on improving this in later PRs.

File: dev/run-tests.py [hunk 3]
```diff
@@ -227,11 +228,32 @@ def build_spark_documentation():
     os.chdir(SPARK_HOME)
 
 
+def get_zinc_port():
+    """
+    Get a randomized port on which to start Zinc
+    """
+    return random.randrange(3030, 4030)
```
Comment (vanzin) [inline dev/run-tests.py]: I hope this doesn't lead to builds failing because zinc can't bind to the chosen port.

Comment (pwendell) [inline dev/run-tests.py]: This logic is identical to that hard coded in the bash scripts that run the maven builds, in the past they've never failed for this reason.

**Comment by pwendell** [comment]
Thanks for looking at this @vanzin. I do agree it would be a lot nicer to base things on comments, but because the comment stream isn't available as meta-data on jenkins, that's a huge amount of additional work that IMO is best left to an extension if someone is feeling interested. Given that it seems like you are cool to merge this for now as is, then look at evolving it more later.

**Pull request closed without merging**